### ISSUE TITLE
Match Error with Play 2.4.0-M2 when Using Cache API

### ISSUE THREAD

**Issue by michaelahlers** [scaldi/scaldi-play#10]
Following from [your Play 2.4.0-M2 integration example](https://github.com/OlegIlyenko/scaldi-play-example) using the `0.5-play2.4-2` release, I get the following error when making web requests to a simple application:

```
! @6l9ii3n53 - Internal server error, for (GET) [/] ->

play.api.UnexpectedException: Unexpected exception[MatchError: QualifierInstance(@play.cache.NamedCache(value=play)) (of class play.api.inject.QualifierInstance)]
	at play.core.ReloadableApplication$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(ApplicationProvider.scala:176) ~[play_2.11-2.4.0-M2.jar:2.4.0-M2]
	at play.core.ReloadableApplication$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(ApplicationProvider.scala:130) ~[play_2.11-2.4.0-M2.jar:2.4.0-M2]
	at scala.Option.map(Option.scala:146) ~[scala-library-2.11.5.jar:na]
	at play.core.ReloadableApplication$$anonfun$get$1$$anonfun$apply$1.apply(ApplicationProvider.scala:130) ~[play_2.11-2.4.0-M2.jar:2.4.0-M2]
	at play.core.ReloadableApplication$$anonfun$get$1$$anonfun$apply$1.apply(ApplicationProvider.scala:128) ~[play_2.11-2.4.0-M2.jar:2.4.0-M2]
Caused by: scala.MatchError: QualifierInstance(@play.cache.NamedCache(value=play)) (of class play.api.inject.QualifierInstance)
	at scaldi.play.ScaldiApplicationLoader$$anonfun$12.apply(ScaldiApplicationLoader.scala:128) ~[scaldi-play_2.11-0.5-play2.4-2.jar:0.5-play2.4-2]
	at scaldi.play.ScaldiApplicationLoader$$anonfun$12.apply(ScaldiApplicationLoader.scala:128) ~[scaldi-play_2.11-0.5-play2.4-2.jar:0.5-play2.4-2]
	at scala.Option.map(Option.scala:146) ~[scala-library-2.11.5.jar:na]
	at scaldi.play.ScaldiApplicationLoader$.identifiersForKey(ScaldiApplicationLoader.scala:128) ~[scaldi-play_2.11-0.5-play2.4-2.jar:0.5-play2.4-2]
	at scaldi.play.ScaldiApplicationLoader$$anonfun$13$$anonfun$apply$2.apply(ScaldiApplicationLoader.scala:144) ~[scaldi-play_2.11-0.5-play2.4-2.jar:0.5-play2.4-2]
```

Does that mean anything to you?

For what it's worth, your exemplar works great, and the only key differences between it and my application are:

1. Nested routes files (_e.g._, `-> /path path.Routes` in `conf/routes`).
2. Bound resources with symbol names (_e.g._, `bind[SomeClass] identifiedBy 'mySymbol to new SomeClass`).

**Comment by michaelahlers** [comment]
It's real easy to trigger this behavior by including Play's Cache API. This behavior also trips unit tests when `"com.typesafe.play" %% "play-cache" % "2.4.0-M2" % "provided"` is added to `build.sbt`.

**Comment by michaelahlers** [comment]
The naïve fix is to add `case QualifierInstance(a: NamedCache) => StringIdentifier(a.value())` to `scaldi.play.ScaldiApplicationLoader.identifiersForKey`, but that intrinsically creates a dependency on `play-cache`, which not every project wants.

**Comment by michaelahlers** [comment]
A concrete instance of [`NamedCache`](https://github.com/playframework/playframework/blob/master/framework/src/play-cache/src/main/java/play/cache/NamedCache.java), [`NamedCacheImpl`](https://github.com/playframework/playframework/blob/master/framework/src/play-cache/src/main/java/play/cache/NamedCacheImpl.java), is passed as the `qualifier` value on `key` to `ScaldiApplicationLoader.identifiersForKey`. A direct fix is to map across its interfaces, get each `Qualifier` annotation, filter out `null`, and check for empty. I'll create a pull request shortly.

**Comment by OlegIlyenko** [comment]
Thanks a lot for reporting it! I briefly looked at the problem and looks like provided  `play.cache.NamedCache` annotation is actually an instance of `play.cache.NamedCacheImpl` class. This looks a little bit strange to me on the first sight, but this causes the `MachError`, which expects the `Qualifier` annotation directly on the provided annotation object.

This kind of use-case is not very clearly specified by JSR-330 spec and by the looks of it not covered by it's TCK.

In the next few days I will try to find a way to fix this issue. Looks like fix should be pretty trivial, but i need to make a little research in order to implement it in JSR-330  compliant way.

**Comment by OlegIlyenko** [comment]
Sorry, haven't seen your last comment before posting. Yeah you are right, it's because of the `NamedCacheImpl`. I think it will require new type of `Identifier` because `AnnotationIdentifier` only capable of storing and comparing the `Type`s of the annotation and not concrete instances of them (and according to the spec, `Qualifier` annotations allowed to have attributes on them). But I'm not sure yet

**Comment by michaelahlers** [comment]
And… I went ahead and issued a (possibly invalid) pull-request before reading your comments. :stuck_out_tongue_winking_eye:

**Comment by OlegIlyenko** [comment]
You are super fast! :) Thanks a lot, I will definitely look at it tomorrow

**Comment by michaelahlers** [comment]
@OlegIlyenko, thank you for your time!

**Comment by OlegIlyenko** [comment]
Issue is fixed. It turns out that it was a little bit more involving, as I expected. Because of this I also released the new version of scaldi:

https://github.com/scaldi/scaldi/releases/tag/v0.5.4

I also created a temporary workaround for this issue:

https://issues.scala-lang.org/browse/SI-9177

As far as I saw you also faced it in your PR. Hopefully it would be resolved soon.

I just released scaldi-play version `0.5-play2.4-3` with all of the fixes. Soon it should be available in the maven central. It would be nice if you could check it out and see whether is solves all of the problems in your app.

**Comment by michaelahlers** [comment]
@OlegIlyenko, your report in [SI-9177](https://issues.scala-lang.org/browse/SI-9177) is exactly what I saw. Also, confirmed that running with `0.5-play2.4-3` addresses my original issue. Great work; I really appreciate it!
